### PR TITLE
Retry SpnegoAuthentication when Network is unreachable

### PR DESCRIPTION
**Pull request by kokosing**
Retry SpnegoAuthentication when Network is unreachable

### CODE REVIEW

File: http-client/src/main/java/io/airlift/http/client/spnego/SpnegoAuthentication.java [hunk 1]
```diff
@@ -147,9 +153,36 @@ public void apply(Request request)
...
+                        .run("getToken", () -> context.initSecContext(new byte[0], 0, 0));
+            }
         };
     }
 
+    private static boolean networkIssue(Throwable exception)
+    {
+        requireNonNull(exception, "exception is null");
+        if (exception == exception.getCause()) {
+            return false;
```
Comment (findepi) [inline http-client/src/main/java/io/airlift/http/client/spnego/SpnegoAuthentication.java]:
```
com.google.common.base.Throwables.getCausalChain(exception).stream()
.anyMatch(exception -> exception instanceof ConnectException &&
  "Network is unreachable".equals(exception.getMessage());
```

File: http-client/src/main/java/io/airlift/http/client/spnego/SpnegoAuthentication.java [hunk 2]
```diff
@@ -147,9 +153,36 @@ public void apply(Request request)
...
+        }
+
+        if (exception instanceof GSSException) {
+            if (exception.getCause() != null) {
+                return networkIssue(exception.getCause());
+            }
+            return false;
+        }
+        if (exception instanceof ConnectException) {
+            return "Network is unreachable".equals(exception.getMessage());
```
Comment (findepi) [inline http-client/src/main/java/io/airlift/http/client/spnego/SpnegoAuthentication.java]: Can we retry on all `ConnectException`?

File: http-client/src/main/java/io/airlift/http/client/spnego/SpnegoAuthentication.java [hunk 3]
```diff
@@ -147,9 +153,36 @@ public void apply(Request request)
                     }
                 }
             }
+
+            private byte[] getToken(GSSContext context)
```
Comment (findepi) [inline http-client/src/main/java/io/airlift/http/client/spnego/SpnegoAuthentication.java]: `initSecContext` ?
(in method name and in `.run(...`)

**Comment by findepi** [comment]
Are you really merging this to `release-0.176-e` ?

File: http-client/src/main/java/io/airlift/http/client/RetryDriver.java [hunk 4]
```diff
@@ -71,34 +72,48 @@ public static RetryDriver retry()
...
+        List<Predicate<Exception>> stopConditions = ImmutableList.<Predicate<Exception>>builder()
+                .addAll(stopCondititions)
+                .addAll(classesAsPredicates)
                 .build();
 
-        return new RetryDriver(maxAttempts, minSleepTime, maxSleepTime, scaleFactor, maxRetryTime, exceptionMapper, exceptions, retryRunnable);
+        return new RetryDriver(maxAttempts, minSleepTime, maxSleepTime, scaleFactor, maxRetryTime, exceptionMapper, stopConditions, retryRunnable);
+    }
+
+    public final RetryDriver stopOn(Predicate<Exception>... stopCondition)
```
Comment (sopel39) [inline http-client/src/main/java/io/airlift/http/client/RetryDriver.java]: nit: in fact `stopCondition` is also `stopConditions` (plural)

File: http-client/src/main/java/io/airlift/http/client/spnego/SpnegoAuthentication.java (hunk 1 above)
Comment (sopel39) [reply http-client/src/main/java/io/airlift/http/client/spnego/SpnegoAuthentication.java]: nice

**Comment by kokosing** [comment]
> Are you really merging this to `release-0.176-e` ?

Just to merge it to our release, I am going to open a PR to airlfit.

**Pull request merged**